File: examples/go/facilitator/advanced/gas_extensions.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"time"

	"github.com/gin-gonic/gin"
	x402 "github.com/x402-foundation/x402/go"
	eip2612gassponsor "github.com/x402-foundation/x402/go/extensions/eip2612gassponsor"
	"github.com/x402-foundation/x402/go/extensions/erc20approvalgassponsor"
	evm "github.com/x402-foundation/x402/go/mechanisms/evm/exact/facilitator"
	uptoevm "github.com/x402-foundation/x402/go/mechanisms/evm/upto/facilitator"
)

/**
 * Gas extensions facilitator example (exact + upto)
 *
 * Registers `exact` and `upto` on Base Sepolia and advertises both Permit2 gas-sponsoring
 * extensions: EIP-2612 and ERC-20 approve (tokens without EIP-2612).
 *
 * Requires EVM_PRIVATE_KEY with Base Sepolia ETH for settlement and for broadcasting
 * client-supplied approval transactions when those extensions are used.
 */

func runGasExtensionsExample(evmPrivateKey string) error {
	evmNetwork := x402.Network("eip155:84532")

	evmSigner, err := newFacilitatorEvmSigner(evmPrivateKey, DefaultEvmRPC)
	if err != nil {
		return fmt.Errorf("failed to create EVM signer: %w", err)
	}

	facilitator := x402.Newx402Facilitator()

	evmConfig := &evm.ExactEvmSchemeConfig{
		DeployERC4337WithEIP6492: true,
	}
	facilitator.Register([]x402.Network{evmNetwork}, evm.NewExactEvmScheme(evmSigner, evmConfig))
	facilitator.Register([]x402.Network{evmNetwork}, uptoevm.NewUptoEvmScheme(evmSigner, nil))

	facilitator.RegisterExtension(eip2612gassponsor.EIP2612GasSponsoring)
	facilitator.RegisterExtension(&erc20approvalgassponsor.Erc20ApprovalFacilitatorExtension{
		Signer: newErc20ApprovalGasSponsorSigner(evmSigner),
	})

	facilitator.OnAfterVerify(func(ctx x402.FacilitatorVerifyResultContext) error {
		fmt.Printf("✅ Payment verified\n")
		return nil
	})

	facilitator.OnAfterSettle(func(ctx x402.FacilitatorSettleResultContext) error {
		fmt.Printf("🎉 Payment settled: %s\n", ctx.Result.Transaction)
		return nil
	})

	gin.SetMode(gin.ReleaseMode)
	r := gin.New()
	r.Use(gin.Recovery())

	r.GET("/supported", func(c *gin.Context) {
		supported := facilitator.GetSupported()
		c.JSON(http.StatusOK, supported)
	})

	r.GET("/health", func(c *gin.Context) {
		c.JSON(http.StatusOK, gin.H{"status": "ok"})
	})

	r.POST("/verify", func(c *gin.Context) {
		ctx, cancel := context.WithTimeout(c.Request.Context(), 30*time.Second)
		defer cancel()

		var reqBody struct {
			PaymentPayload      json.RawMessage `json:"paymentPayload"`
			PaymentRequirements json.RawMessage `json:"paymentRequirements"`
		}

		if err := c.BindJSON(&reqBody); err != nil {
			c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid request body"})
			return
		}

		result, err := facilitator.Verify(ctx, reqBody.PaymentPayload, reqBody.PaymentRequirements)
		if err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
			return
		}

		c.JSON(http.StatusOK, result)
	})

	r.POST("/settle", func(c *gin.Context) {
		ctx, cancel := context.WithTimeout(c.Request.Context(), 60*time.Second)
		defer cancel()

		var reqBody struct {
			PaymentPayload      json.RawMessage `json:"paymentPayload"`
			PaymentRequirements json.RawMessage `json:"paymentRequirements"`
		}

		if err := c.BindJSON(&reqBody); err != nil {
			c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid request body"})
			return
		}

		result, err := facilitator.Settle(ctx, reqBody.PaymentPayload, reqBody.PaymentRequirements)
		if err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
			return
		}

		c.JSON(http.StatusOK, result)
	})

	fmt.Printf("🚀 Gas extensions facilitator (exact + upto) listening on http://localhost:%s\n", defaultPort)
	fmt.Printf("   Extensions: eip2612GasSponsoring, erc20ApprovalGasSponsoring\n")
	fmt.Printf("   EVM: %s on %s\n", evmSigner.GetAddresses()[0], evmNetwork)
	fmt.Println()

	return r.Run(":" + defaultPort)
}

File: examples/typescript/facilitator/advanced/gas_extensions.ts
/**
 * Permit2 gas sponsorship facilitator example (`exact` + `upto`)
 *
 * Registers both `exact` and `upto` EVM schemes on Base Sepolia and advertises the two
 * gas-sponsoring extensions: EIP-2612 permits and ERC-20 `approve` transactions (tokens
 * without EIP-2612). Both schemes use Permit2 paths that can pair with these extensions.
 *
 * Requires `EVM_PRIVATE_KEY` with Base Sepolia ETH for settlement and for broadcasting
 * client-supplied approval transactions when those extensions are used.
 */

import { x402Facilitator } from "@x402/core/facilitator";
import {
  PaymentPayload,
  PaymentRequirements,
  SettleResponse,
  VerifyResponse,
} from "@x402/core/types";
import { toFacilitatorEvmSigner } from "@x402/evm";
import { ExactEvmScheme } from "@x402/evm/exact/facilitator";
import { UptoEvmScheme } from "@x402/evm/upto/facilitator";
import {
  EIP2612_GAS_SPONSORING,
  createErc20ApprovalGasSponsoringExtension,
} from "@x402/extensions";
import dotenv from "dotenv";
import express from "express";
import { createWalletClient, http, publicActions } from "viem";
import { privateKeyToAccount } from "viem/accounts";
import { baseSepolia } from "viem/chains";

dotenv.config();

const PORT = process.env.PORT || "4022";
const EVM_NETWORK = "eip155:84532";

if (!process.env.EVM_PRIVATE_KEY) {
  console.error("❌ EVM_PRIVATE_KEY environment variable is required");
  process.exit(1);
}

const evmAccount = privateKeyToAccount(
  process.env.EVM_PRIVATE_KEY as `0x${string}`,
);
console.info(`EVM Facilitator account: ${evmAccount.address}`);

const viemClient = createWalletClient({
  account: evmAccount,
  chain: baseSepolia,
  transport: http(),
}).extend(publicActions);

const evmSigner = toFacilitatorEvmSigner({
  getCode: (args: { address: `0x${string}` }) => viemClient.getCode(args),
  address: evmAccount.address,
  readContract: (args: {
    address: `0x${string}`;
    abi: readonly unknown[];
    functionName: string;
    args?: readonly unknown[];
  }) =>
    viemClient.readContract({
      ...args,
      args: args.args || [],
    }),
  verifyTypedData: (args: {
    address: `0x${string}`;
    domain: Record<string, unknown>;
    types: Record<string, unknown>;
    primaryType: string;
    message: Record<string, unknown>;
    signature: `0x${string}`;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
  }) => viemClient.verifyTypedData(args as any),
  writeContract: (args: {
    address: `0x${string}`;
    abi: readonly unknown[];
    functionName: string;
    args: readonly unknown[];
  }) =>
    viemClient.writeContract({
      ...args,
      args: args.args || [],
    }),
  sendTransaction: (args: { to: `0x${string}`; data: `0x${string}` }) =>
    viemClient.sendTransaction(args),
  waitForTransactionReceipt: (args: { hash: `0x${string}` }) =>
    viemClient.waitForTransactionReceipt(args),
});

const facilitator = new x402Facilitator()
  .onBeforeVerify(async (context) => {
    console.log("Before verify", context);
  })
  .onAfterVerify(async (context) => {
    console.log("After verify", context);
  })
  .onVerifyFailure(async (context) => {
    console.log("Verify failure", context);
  })
  .onBeforeSettle(async (context) => {
    console.log("Before settle", context);
  })
  .onAfterSettle(async (context) => {
    console.log("After settle", context);
  })
  .onSettleFailure(async (context) => {
    console.log("Settle failure", context);
  });

facilitator.register(
  EVM_NETWORK,
  new ExactEvmScheme(evmSigner, { deployERC4337WithEIP6492: true }),
);
facilitator.register(EVM_NETWORK, new UptoEvmScheme(evmSigner));

const erc20ApprovalSigner = {
  ...evmSigner,
  sendTransactions: async (
    transactions: (
      | `0x${string}`
      | { to: `0x${string}`; data: `0x${string}`; gas?: bigint }
    )[],
  ): Promise<`0x${string}`[]> => {
    const hashes: `0x${string}`[] = [];
    for (const tx of transactions) {
      let hash: `0x${string}`;
      if (typeof tx === "string") {
        hash = await viemClient.sendRawTransaction({
          serializedTransaction: tx,
        });
      } else {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        hash = await viemClient.sendTransaction(tx as any);
      }
      const receipt = await viemClient.waitForTransactionReceipt({ hash });
      if (receipt.status !== "success") {
        throw new Error(`transaction_failed: ${hash}`);
      }
      hashes.push(hash);
    }
    return hashes;
  },
};

facilitator
  .registerExtension(EIP2612_GAS_SPONSORING)
  .registerExtension(
    createErc20ApprovalGasSponsoringExtension(erc20ApprovalSigner),
  );

const app = express();
app.use(express.json());

app.post("/verify", async (req, res) => {
  try {
    const { paymentPayload, paymentRequirements } = req.body as {
      paymentPayload: PaymentPayload;
      paymentRequirements: PaymentRequirements;
    };

    if (!paymentPayload || !paymentRequirements) {
      return res.status(400).json({
        error: "Missing paymentPayload or paymentRequirements",
      });
    }

    const response: VerifyResponse = await facilitator.verify(
      paymentPayload,
      paymentRequirements,
    );

    res.json(response);
  } catch (error) {
    console.error("Verify error:", error);
    res.status(500).json({
      error: error instanceof Error ? error.message : "Unknown error",
    });
  }
});

app.post("/settle", async (req, res) => {
  try {
    const { paymentPayload, paymentRequirements } = req.body;

    if (!paymentPayload || !paymentRequirements) {
      return res.status(400).json({
        error: "Missing paymentPayload or paymentRequirements",
      });
    }

    const response: SettleResponse = await facilitator.settle(
      paymentPayload as PaymentPayload,
      paymentRequirements as PaymentRequirements,
    );

    res.json(response);
  } catch (error) {
    console.error("Settle error:", error);

    if (
      error instanceof Error &&
      error.message.includes("Settlement aborted:")
    ) {
      return res.json({
        success: false,
        errorReason: error.message.replace("Settlement aborted: ", ""),
        network: req.body?.paymentPayload?.network || "unknown",
      } as SettleResponse);
    }

    res.status(500).json({
      error: error instanceof Error ? error.message : "Unknown error",
    });
  }
});

app.get("/supported", async (req, res) => {
  try {
    const response = facilitator.getSupported();
    res.json(response);
  } catch (error) {
    console.error("Supported error:", error);
    res.status(500).json({
      error: error instanceof Error ? error.message : "Unknown error",
    });
  }
});

app.get("/health", (req, res) => {
  res.json({ status: "ok" });
});

app.listen(parseInt(PORT), () => {
  console.log(
    `🚀 Gas extensions facilitator (exact + upto) listening on http://localhost:${PORT}`,
  );
  console.log(
    `   Extensions: eip2612GasSponsoring, erc20ApprovalGasSponsoring`,
  );
  console.log();
});
